pkg/provisioner/crdprovisioner.go

```diff
@@ -251,37 +257,64 @@ func (c *CrdProvisioner) CreateVolume(
 		}()
 		return nil, err
 	}
+	azVolumeInstance = obj.(*v1alpha1.AzVolume)
 
 	return azVolumeInstance.Status.Detail.ResponseObject, nil
```

Comment:
If `obj == nil && err == nil`, the type assertion above will produce `nil` and this line will result in a `nil` dereference. This may explain the [e2e test failure](https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/runs/3738929638).

Reply:
This should be handled by the if clause above:

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/a4c58f8546f8240542c248db08debbf5254fc2fd/pkg/provisioner/crdprovisioner.go#L244

Reply:
this should be handled by the if clause above
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/a4c58f8546f8240542c248db08debbf5254fc2fd/pkg/provisioner/crdprovisioner.go#L244

Reply:
Ah, yes, my mistake. Still it seems some condition results in a 'nil' dereference on line 262. I'm guessing `Detail` was `nil` then.

```
•! Panic [1.379 seconds]
Pre-Provisioned
/home/runner/work/azuredisk-csi-driver/azuredisk-csi-driver/test/e2e/pre_provisioning_test.go:37
  [single-az]
  /home/runner/work/azuredisk-csi-driver/azuredisk-csi-driver/test/e2e/pre_provisioning_test.go:70
    should fail when maxShares is invalid [disk.csi.azure.com][windows] [It]
    /home/runner/work/azuredisk-csi-driver/azuredisk-csi-driver/test/e2e/pre_provisioning_test.go:302

    Test Panicked
    runtime error: invalid memory address or nil pointer dereference
    /opt/hostedtoolcache/go/1.17.1/x64/src/runtime/panic.go:221

    Full Stack Trace
    sigs.k8s.io/azuredisk-csi-driver/pkg/provisioner.(*CrdProvisioner).CreateVolume(0xc000937740, {0x2311e08, 0xc000056100}, {0x1fbed66, 0x12}, 0xc00066e480, {0xc000b5b400, 0x1, 0x1}, 0xc000941ce0, ...)
    	/home/runner/work/azuredisk-csi-driver/azuredisk-csi-driver/pkg/provisioner/crdprovisioner.go:262 +0xaf0
    sigs.k8s.io/azuredisk-csi-driver/pkg/azuredisk.(*DriverV2).CreateVolume(0xc000653380, {0x2311e08, 0xc000056100}, 0xc00001e310)
    	/home/runner/work/azuredisk-csi-driver/azuredisk-csi-driver/pkg/azuredisk/controllerserver_v2.go:120 +0xc74
    sigs.k8s.io/azuredisk-csi-driver/test/e2e.glob..func4.3.7()
    	/home/runner/work/azuredisk-csi-driver/azuredisk-csi-driver/test/e2e/pre_provisioning_test.go:310 +0x24f
    sigs.k8s.io/azuredisk-csi-driver/test/e2e.TestE2E(0x407c59)
    	/home/runner/work/azuredisk-csi-driver/azuredisk-csi-driver/test/e2e/suite_test.go:247 +0x178
    testing.tRunner(0xc0006529c0, 0x20ab1d0)
    	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:1259 +0x102
    created by testing.(*T).Run
    	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:1306 +0x35a
```

Reply:
Ah yes, that probably is the case.

The problem seems to be that the condition function is returning `true, err` when error is not nil.

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/a4c58f8546f8240542c248db08debbf5254fc2fd/pkg/provisioner/crdprovisioner.go#L155-L157

I should probably check for error first in `conditionwatcher`
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/a4c58f8546f8240542c248db08debbf5254fc2fd/pkg/provisioner/conditionwatcher.go#L149-L154